feat: add a prototype for the Guild Loot Tracker workshop

Sebastian-Wlo opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a [descriptive title](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request?id=prepare-a-good-pr-title) (**not** a vague title like `Update index.md`)

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Relates to: https://github.com/freeCodeCamp/freeCodeCamp/issues/64119

<!-- Feel free to add any additional description of changes below this line -->
- link to the related issue: https://github.com/freeCodeCamp/freeCodeCamp/issues/64119 ,
- currently, the prototype uses `for()` loops and `Array.prototype.sort()` method,
- I couldn't figure out how to use "Object.values()" in the functions (maybe adding a function returning the sums of ingredient macros to the requirements would be a good place to implement it?)

Sebastian-Wlo committed:
feat: add prototype for Nutrient Tracker Dictionary lab prototype

This adds a prototype for the Nutrient Tracker Dictionary lab.
Currently, the prototype uses "for()" loops and ".sort()" method.
Couldn't figure out how to use "Object.values()" in the functions.

jdwilkin4 reviewed (commented):
Please make to include the user stories so those can be review too just in case updates need to be made there. 

Also, please let us know where this falls in the curriuclum.

loops module?
review modules?
higher order functions?

 we will need that information when conducting reviews 👍🏾

Sebastian-Wlo commented:
Just to make sure: should the user stories be a copy of those mentioned in the related issue, or should I come up with some new ones?

As for where this lab should be placed in the curriculum - originally, it was supposed to be in the Objects module. But I think `listTopIngredients()`, which was supposed to return sorted output requires the use of a loop. `sort()` would be ideal, but it should be doable with just a `for()` loop (I'd definitely need to make some changes to make it work).

So, Loops would probably be the right place for it.

Thank you, I'll add the user stories as soon as I can.

Sebastian-Wlo committed:
feat: add user-stories.md to the nutrient tracker prototype

majestic-owl448 commented:
there are many other things that could be tracked in a pretty similar way, so the prototype is still good, there will be need to change the name of some things before it's ready tho

Sebastian-Wlo committed:
chore: renamed "lab-nutrient-tracker-dictionary" directory

Changed "lab-nutrient-tracker-dictionary" to "lab-guild-loot-tracker" to fit the new topic for the Lab.

Sebastian-Wlo committed:
docs: change "nutrient tracker" references in lab's "user-stories.md"

Changes all references in "user-stories.md" from nutrient tracking to guild loot management, and "dictionaries" to "objects".

Sebastian-Wlo committed:
refactor: rewrite of "nutrient tracker" functions

Changed references "nutrient tracking" references in "script.js" to tracking guild loot management.

jdwilkin4 reviewed (commented):
I left some thoughts on the user stories.
But also noticed that none of these functions are being used anywhere. 

From an educational standpoint, we don't want to send the message to beginners to create functions in their code and not use it. All of these functions should be used in some way so campers can connect the dots and see what is actually happening in the code. Even if it is just logging these function call results to the console.

Sebastian-Wlo committed:
docs: changed the lab's summary section to be more descriptive

jdwilkin4 reviewed (commented):
I can't find where Object.keys, Object.values, and Object.entries is taught in the JS curriculum.

A new lesson can be added to the JS fundamentals module for that. 
But until that can happen, we won't be able to add this lab to the curriculum because we can't ask campers to build projects on topics they haven't learned yet. And this lab seems to be solely dedicated to practicing those methods mentioned earlier. 

So you will need to reach out to Naomi on next steps. Either we are going to wait until the new lesson has been added so this lab can be added. Or not add this new lab at all.

Sebastian-Wlo commented:
> But until that can happen, we won't be able to add this lab to the curriculum because we can't ask campers to build projects on topics they haven't learned yet. And this lab seems to be solely dedicated to practicing those methods mentioned earlier.

That's fair - I've been trying to make the prototype work without these methods, and while it could be done, it felt like it wasn't introducing anything that wasn't already shown in the labs for the loops. It also felt more "awkward" than it probably should, if that makes sense, but I wasn't sure if that was because the lab's description didn't really fit that section, or if it was just me doing a bad job of it.

> So you will need to reach out to Naomi on next steps. Either we are going to wait until the new lesson has been added so this lab can be added. Or not add this new lab at all.

I'll do that, thank you!

jdwilkin4 commented:
@Sebastian-Wlo 

What's the status for this lab?

Sebastian-Wlo committed:
refactor: change "listTopMembers" function to "listMembers"

Removed the "listTopMembers" function and added "listMembers", which logs the input object entries to the console.

Sebastian-Wlo committed:
docs: remove the "listTopMembers" mentions from "user-stories.md", add description of "listMembers"

Replaces description of the removed "listTopMembers" function with one for "listMembers".

majestic-owl448 commented:
at the prototype stage, the main difference is that a workshop does not need user stories, and that the protytpe is the final solution of the project, not just a possible solution

jdwilkin4 commented:
@Sebastian-Wlo 

Can you post any updates you have regarding this project?
Thanks 👍🏾

Sebastian-Wlo commented:
@jdwilkin4, Sorry for the lack of any activity!
I know that this will get turned into a workshop - so, I need to remove the remaining `for()` loop, make sure that the prototype is cleanly written and get rid of the user stories before planning out how to break the code into steps.
I'll do my best to update the prototype over the weekend, or in the worst case early next week.

Sebastian-Wlo committed:
refactor: removed users stories for Guild Loot Tracker

This removes user stories from the prototype - Guild Loot Tracker is supposed to be a workshop, and workshops don't use them.

Sebastian-Wlo committed:
refactor: restructured Guild Loot Tracker prototype

This changes the structure of the Guild Loot Tracker to fit better as the workshop.

Sebastian-Wlo committed:
chore: remove the "lab-" prefix from the prototype's directory name

This renames the directory the code is contained in.
It's done to avoid potential confusion, since the purpose of this prototype changed from a Lab to a Workshop.

Sebastian-Wlo committed:
feat: add listMemberNames() function that introduces Object.keys() method

Sebastian-Wlo committed:
refactor: rewrite getMemberTotal() to return an object or `false`

This changes the getMemberTotals() so it either return an object including the member's resources or `false` if the guild object doesn't include that member.

Sebastian-Wlo committed:
refactor: rewrite 'listMembers', 'getMemberTotals', and 'addLootEntry', add coments

This changes the way the "main" functions of the script work., and adds comments to make describing the workshop's steps easier.

Sebastian-Wlo commented:
Terribly sorry it took so long! I've unfortunately been really busy.

Sorry the prototype is not up to an acceptable level. Aside from from all of the (too many) `TODO`s in the current version, are there any glaring mistakes that need to be addressed?

During the conversation in the Naomi's sprints channel it turned out that it's not clear the names I've used for the guild members are actually names (which I should have seen coming, to be honest).
I was thinking about using the names of the fCC staff, like it's been done in some of the previous workshops, but it didn't feel comfortable with assigning people reputation and experience arbitrarily. Would it be okay if I did that? I don't want anyone to be offended about the "stats" I'm assigning here

jdwilkin4 commented on fullstack-cert/js-projects/guild-loot-tracker/script.js:
```diff
@@ -0,0 +1,157 @@
+// This will be a workshop for the Loops section, probably after the "Build a Profile Lookup" lab.
+ 
+// Creating the "guild" object could be 2 or 3 Steps (creating "guild" object itself and adding the first "guild member" object as the first property - rest of the "guild members" can be filled out somewhere between the next steps to avoid too much repetition).
+let guild = {
```
these names should work

```js
let guild = {
  ethan: {
    gold: 31,
    silver: 48,
    reputation: 9,
    experience: 198,
  },
  elara: {
    gold: 78,
    silver: 64,
    reputation: 12,
    experience: 111,
  },
  brandon: {
    gold: 41,
    silver: 7,
    reputation: 7,
    experience: 70,
  },
  dylan: {
    gold: 81,
    silver: 2,
    reputation: 20,
    experience: 220,
  },
  lucas: {
    gold: 34,
    silver: 28,
    reputation: 10,
    experience: 179,
  },
  natalie: {
    gold: 36,
    silver: 81,
    reputation: 12,
    experience: 82,
  },
};
```

jdwilkin4 commented:
> I was thinking about using the names of the fCC staff, like it's been done in some of the previous workshops, but it didn't feel comfortable with assigning people reputation and experience arbitrarily. Would it be okay if I did that? I don't want anyone to be offended about the "stats" I'm assigning here

I left a comment regarding the names. But in general, some of the staff would probably be fine having their names in the curriculum while others might not. So using fake yet more realistic names should be fine here.

Sebastian-Wlo committed:
refactor: change the guild members' names

This changes the names used in the workshop to less cunfusing ones (Thank you Jessica for the list!).

Sebastian-Wlo committed:
refactor: changes the loop in listMemberNames function to 'for...in' for consistency

This changes the 'for' loop in listMemberNames() to a 'for...in' loop to keep the used loops consistent, and adds the use of 'parseInt' to that loop.
Also, adds missing semicolons to listMembers() function (probably should've made it into a separate commit).

Sebastian-Wlo committed:
refactor: change loops in addLootEntry() to 'for...in', fix the order of checks

This changes the loop used for iteration over guild member's names to 'for...in' loop (for consistency), and moves the check if the new entry before updating the member data.

Sebastian-Wlo committed:
feat: use 'getMemberTotals()' function to introduce 'Object.entries()' method

Sebastian-Wlo committed:
refactor: use 'listMembers()' to introduce only the 'Object.values()' method

Sebastian-Wlo committed:
style: format the script for consistency and readability

Sebastian-Wlo commented:
Thank you @jdwilkin4 , that list solves my problem with names!
I still need to clean up the comments and note how to divide the code into steps (I'll try to do that within the next day or two), but **most** of the current version is hopefully ready for the review.

One thing I was thinking about is getting rid of the `listMemberNames()` function and introducing both `Object.keys()` and `Object.values()` in the `listMembers()`.
I'll stick to making any suggested changes (instead of rewriting functions because I randomly came up with a better solution), but I'd like to know if removing `listMemberNames()` is a good idea, or if introducing these two methods in separate functions would be a cleaner way to present them.

jdwilkin4 commented:
> but I'd like to know if removing listMemberNames() is a good idea, or if introducing these two methods in separate functions would be a cleaner way to present them.

We can try the separate functions and then through the review process if we decide it doesn't work as well, we go back to the `listMemberNames`.  👍🏾

Sebastian-Wlo committed:
chore: update comments to reflect steps of the workshop

This updates the overview of how "Guild Loot Tracker" workshop should be broken down into steps.

Sebastian-Wlo commented:
Once again, sorry it took so long - I was too optimistic about how much time I'd have this week... On the upside, I _think_ all of the changes I wanted to make are done, and hopefully the steps breakdown is reasonably clear.
I don't expect the prototype to pass the review on the first go, but hopefully some of it will be usable:)

jdwilkin4 reviewed (commented):
After reviewing this and seeing this will be a workshop after the profile lookup lab, I would suggest simplifying this and keeping the core learning concepts to working with Object.keys(), Object.values(), and Object.entries()

Here is my proposed update

```js
const guild = {
  ethan: {
    gold: 31,
    reputation: 9,
  },
  elara: {
    gold: 78,
    reputation: 12,
  },
  brandon: {
    gold: 41,
    reputation: 7,
  },
  dylan: {
    gold: 81,
    reputation: 20,
  },
};

function listMembers(guildObject) {
  console.log("Guild Members:");

  const members = Object.keys(guildObject);

  for (const member of members) {
    console.log(member);
  }
}

listMembers(guild);

console.log("------------");

function countMembers(guildObject) {
  return Object.keys(guildObject).length;
}

console.log(`Guild Members: ${countMembers(guild)}`);

console.log("------------");

function listGold(guildObject) {
  console.log("Gold Totals:");

  for (const [name, stats] of Object.entries(guildObject)) {
    console.log(`${name}: ${stats.gold} gold`);
  }
}

listGold(guild);

console.log("------------");

function getTotalGold(guildObject) {
  let totalGold = 0;

  for (const member of Object.values(guildObject)) {
    totalGold += member.gold;
  }

  return totalGold;
}

console.log(`Total Gold: ${getTotalGold(guild)}`);

console.log("------------");

function getRichestMember(guildObject) {
  let richestName = "";
  let highestGold = 0;

  for (const [name, stats] of Object.entries(guildObject)) {
    if (stats.gold > highestGold) {
      highestGold = stats.gold;
      richestName = name;
    }
  }

  return richestName;
}

console.log(`Richest Member: ${getRichestMember(guild)}`);

console.log("------------");

function listVeterans(guildObject) {
  console.log("Veteran Members:");

  for (const [name, stats] of Object.entries(guildObject)) {
    if (stats.reputation >= 10) {
      console.log(name);
    }
  }
}

listVeterans(guild);

console.log("------------");

function awardGold(guildObject, amount) {
  for (const member of Object.values(guildObject)) {
    member.gold += amount;
  }
}

awardGold(guild, 10);

console.log("After Awarding 10 Gold:");

listGold(guild);

console.log("------------");

function guildReport(guildObject) {
  console.log("Guild Report");
  console.log("------------");
  console.log(`Members: ${countMembers(guildObject)}`);
  console.log(`Total Gold: ${getTotalGold(guildObject)}`);
  console.log(`Richest Member: ${getRichestMember(guildObject)}`);
}

guildReport(guild);

```

Sebastian-Wlo commented:
That's quite a departure from the current version, but it makes sense considering the whole idea was moved to a different section, as well as from a lab to a workshop:)
It's a shame that object cloning didn't make the cut,  but it makes sense to keep the scope of the workshop more focused on a specific subject. 

I like how much cleaner it looks! But there are 2 things I'd like to ask about before rewriting it:
1. Should `reputation` be left without any function modifying the values? It's fine, it's an exercise and not an actual management app after all (and it keeps this workshop less repetitive), but it does look a little odd to have a value and no way to change it,
2. Would it be ok to separate the "headers" of the function responses (so, "Guild Members:", "Gold Totals:" etc) from the rest of the output with `------------` and functions with just `console.log("")`? Either that, or removing the lines altogether - they look nice, but it's easy to get the number of characters wrong, so maybe it would be better to avoid adding them in multiple places

jdwilkin4 commented:
> Should reputation be left without any function modifying the values? It's fine, it's an exercise and not an actual management app after all (and it keeps this workshop less repetitive), but it does look a little odd to have a value and no way to change it,

I think it is fine to leave reputation. 

> Either that, or removing the lines altogether - they look nice, but it's easy to get the number of characters wrong, so maybe it would be better to avoid adding them in multiple places

If you want to remove them, that is fine

majestic-owl448 commented:
@Sebastian-Wlo do you have any update on this prototype?

Sebastian-Wlo commented:
@majestic-owl448 yes, sorry, I've been very busy recently. I have started re-writing it as per Jessica's proposed update (with added step breakdown) and will try to push the changes in the next 48 hours.

majestic-owl448 commented:
it looks like this is intended to go in the objects module, right?

if that so, keep in mind loops are introduced only later, if it goes in loops then everything ok

<img width="771" height="187" alt="image" src="https://github.com/user-attachments/assets/6e22cc4a-0152-4b8c-93fe-70697e04321a" />

Sebastian-Wlo commented:
From what I understand it should end up in the "Loops" section, somewhere after the "Build a Profile Lookup" Lab - which, on one hand does have a lot of steps already, but on the other has a lot of Labs back to back. I hope that's ok

majestic-owl448 commented:
yes, it's fine, it also has practice of something we don't have much of

Sebastian-Wlo committed:
feat: reworked Guild Loot Tracker workshop according to recommendations

This changes (simplifies) the workshop in accordance to @jdwilkin4 proposed update.

Sebastian-Wlo commented:
I started writing instructions for each step, and the current version can get a bit to packed at times. Not a problem with the code itself, but the step breakdown in the comments could be better.
The final version shouldn't end up with more than 35 steps, but it's still more than currently indicated.

I'm not sure if I should fix the comments now, since I'd imagine that when I get to actually breaking the prototype into steps there will be some other minor changes to the pacing anyway.

jdwilkin4 commented:
> I'm not sure if I should fix the comments now, since I'd imagine that when I get to actually breaking the prototype into steps there will be some other minor changes to the pacing anyway.

I wouldn't worry to much about the steps and descriptions at this stage. 
The goal of this stage is to have a working prototype that the maintainers can review that works in the current place in the curriculum.

Once the prototype is finalized, then you can worry about steps and tests. Also, in the next review stage in the main repo, maintainers will probably have suggestions for breaking things down further if needed.

jdwilkin4 commented:
@Sebastian-Wlo 

If this is ready for review, then we can take another look at it.
If you are still working on this, then continue to push up commits and then let use know when this is ready for review.

Sebastian-Wlo commented:
@jdwilkin4 It is ready for another review - I can't think of anything that would improve the code at this point (it is almost exactly your simplified version of the workshop line per line - also thank you again for writing it! It's definitely better suited to be turned into a Workshop than my previous attempts).
I actually forgot GitHub has a 'Re-request a review' option.

imsomilg commented:
I think we should have `console.log("------------");` instead of `console.log("")`. Maybe we can create a function and keep reusing it so that users don't have to count no. of dashes every time?

majestic-owl448 commented on fullstack-cert/js-projects/guild-loot-tracker/script.js:
```diff
@@ -0,0 +1,172 @@
+
+console.log(`Total gold: ${getTotalGold(guild)}`);
+
+console.log("");
```
```suggestion
console.log(`Total gold: ${getTotalGold(guild)}\n`);
```

majestic-owl448 commented on fullstack-cert/js-projects/guild-loot-tracker/script.js:
```diff
@@ -0,0 +1,172 @@
+
+console.log(`Richest Member: ${getRichestMember(guild)}`);
+
+console.log("");
```
```suggestion
console.log(`Richest Member: ${getRichestMember(guild)}\n`);
```

majestic-owl448 reviewed (commented):
I don't really like having `console.log("");`, it does the same as `console.log()`, and can be replaced at least in some places with a `\n` in the previous step

(it would also make for a boring step, to add an empty console.log, tbh)

(yeah, it's a nitpick, I would say it's ready)

Sebastian-Wlo commented:
Both are good solutions - thank you for your suggestions! I think adding a function for logging a "separator line" would work slightly better, since it will not require any workarounds for functions used in `guildReport()` to look uniform.

Honestly, I went with logging an empty string instead of adding `\n` mainly because asking for a newline character could look a bit confusing in the instructions. Adding it at the end of `"Guild Report"` shouldn't be as much of a problem though, considering it can be given in the step as a whole string.
It would also allow to differentiate the line below the "header" in the `guildReport` and separators between the earlier functions.

Sebastian-Wlo committed:
feat: added 'logSeparator' function to the Guild Loot Tracker workshop

This adds a 'logSeparator' function which logs a separating line to consle (for output readability).
It also replaces the `console.log("")` calls  with the `logSeparator()` calls.

Sebastian-Wlo committed:
refactor: replace the separating line with newline char in Guild Loot Tracker workshop

This replaces the line separating the "header" from the rest of the lines in the `guildReport` function console output.

jdwilkin4 reviewed (approved):
I think these changes look good.
